src/lib.rs

```diff
@@ -546,27 +581,43 @@ pub unsafe trait FromBytes: FromZeroes {
 ///
 /// # Safety
 ///
-/// If `T: AsBytes`, then unsafe code may assume that it is sound to treat any
-/// instance of the type as an immutable `[u8]` of length `size_of::<T>()`. If a
-/// type is marked as `AsBytes` which violates this contract, it may cause
+/// *This section describes what is required in order for `T: AsBytes`, and what
+/// unsafe code may assume of such types. `#[derive(AsBytes)]` only permits
+/// types which satisfy these requirements. If you don't plan on implementing
+/// `AsBytes` manually, and you don't plan on writing unsafe code that operates
+/// on `AsBytes` types, then you don't need to read this section.*
+///
+/// If `T: AsBytes`, then unsafe code may assume that:
+/// - ...it is sound to treat any `t: T` as an immutable `[u8]` of length
+///   `size_of_val(t)`.
+/// - ...given `t: &T`, it is sound to construct a `b: &[u8]` where `b.len() ==
+///   size_of_val(t)` at the same address as `t`, and it is sound for both `b`
+///   and `t` to be live at the same time.
+///
+/// If a type is marked as `AsBytes` which violates this contract, it may cause
 /// undefined behavior.
 ///
-/// If a type has the following properties, then it is safe to implement
-/// `AsBytes` for that type
+/// If a type has the following properties, then it is sound to implement
+/// `AsBytes` for that type:
 /// - If the type is a struct:
 ///   - It must have a defined representation (`repr(C)`, `repr(transparent)`,
 ///     or `repr(packed)`).
-///   - All of its fields must be `AsBytes`
+///   - All of its fields must satisfy the requirements to be `AsBytes` (they do
+///     not actually have to be `AsBytes`).
 ///   - Its layout must have no padding. This is always true for
 ///     `repr(transparent)` and `repr(packed)`. For `repr(C)`, see the layout
 ///     algorithm described in the [Rust Reference].
 /// - If the type is an enum:
-///   - It must be a C-like enum (meaning that all variants have no fields)
+///   - It must be a C-like enum (meaning that all variants have no fields).
```

Comment:
This is also more strict than necessary. It needs to have no padding, which can be guaranteed with some enums with fields, such as:

```rust
#[repr(u32)]
enum Foo {
  A(i32),
  B([u8; 4]),
}
```

If the type is an enum with fields, it must either:
  - Have an explicit representation `#[repr(Int)]`. Each field must be the same size. No field data can have alignment greater than `Int`. All field data must be `AsBytes`.
  - Have a single variant with data that is `AsBytes`. All variants without data are represented as an illegal representation (a niche) of the data type, such that the size of the enum is the size of the data in the single variant. This covers `Option<NonZeroUsize>`. (this could be phrased more cleanly)

I believe both of these could be `derive`d.

Reply:
Filed https://github.com/google/zerocopy/issues/288 to track this.